File: ai_remixmate_feature_lab/python/ai_remixmate_features/intelligence/track_match.py
from __future__ import annotations

from typing import Any

from .compatibility import compatibility_score


def rank_tracks(source: dict[str, Any], candidates: list[dict[str, Any]]) -> list[dict[str, Any]]:
    """Rank candidate tracks by deterministic compatibility score."""
    results = []
    for candidate in candidates:
        if candidate.get("id") == source.get("id"):
            continue
        breakdown = compatibility_score(source, candidate)
        results.append({"trackId": candidate["id"], "score": breakdown["overall"], "breakdown": breakdown})
    return sorted(results, key=lambda item: item["score"], reverse=True)

File: ai_remixmate_feature_lab/typescript/src/intelligence/trackMatch.ts
import type { TrackAnalysis } from "../models/TrackAnalysis";
import type { CompatibilityScore } from "../models/CompatibilityScore";
import { compatibilityScore } from "./compatibilityScore";

export interface TrackMatchResult {
  trackId: string;
  score: number;
  breakdown: CompatibilityScore;
}

export function trackMatch(source: TrackAnalysis, candidates: TrackAnalysis[]): TrackMatchResult[] {
  return candidates
    .filter((candidate) => candidate.id !== source.id)
    .map((candidate) => {
      const breakdown = compatibilityScore(source, candidate);
      return { trackId: candidate.id, score: breakdown.overall, breakdown };
    })
    .sort((a, b) => b.score - a.score);
}
